### PR TITLE
Reorganize batching

### PR DESCRIPTION
**Pull request by cthoyt**
# Summary
 
Currently, there are a few classes that mediate generation of DrugPairBatch objects. This PR does a big overhaul to simplify that code. Most importantly, it tries to shy away from the java-style coding of creating an object then immediately after putting stuff inside it (now just make the object with the stuff in the first place)
 
- [x] Code passes all tests
- [x] Unit tests provided for these changes
- [x] Documentation and docstrings added for these changes

## Changes 

* Turn DrugPairBatch into an actual dataclass
* Create BatchGenerator directly from the dataset (and combine with splitting)

## Next Steps

After this PR is merged, the next steps are to greatly simplify the context and drug feature set classes as well as impose mypy testing (currently the way they subclass dictionary is problematic)

### CODE REVIEW

File: chemicalx/data/batchgenerator.py [hunk 1]
```diff
@@ -149,18 +120,16 @@ def generate_batch(self, batch_frame: pd.DataFrame):
...
-            context_features,
-            labels,
+        return DrugPairBatch(
+            identifiers=batch_frame,
+            drug_features_left=drug_features_left,
+            drug_molecules_left=drug_molecules_left,
+            drug_features_right=drug_features_right,
+            drug_molecules_right=drug_molecules_right,
+            context_features=context_features,
+            labels=labels,
```
Comment (cthoyt) [inline chemicalx/data/batchgenerator.py:130]: Is there ever a time where `labels=False`? Perhaps this was a premature abstraction - I think we can actually remove this argument everywhere.

File: chemicalx/data/drugpairbatch.py [hunk 2]
```diff
@@ -1,38 +1,32 @@
...
+import pandas as pd
+import torch
+from torchdrug.data import PackedGraph
+
 __all__ = [
     "DrugPairBatch",
 ]
 
 
+@dataclass
```
Comment (benedekrozemberczki) [inline chemicalx/data/drugpairbatch.py:15]: This is great!

File: chemicalx/data/batchgenerator.py (hunk 1 above)
Comment (benedekrozemberczki) [reply chemicalx/data/batchgenerator.py:130]: I suppose some unsupervised learning could be a case.

**Review by benedekrozemberczki — APPROVED** [review]

**Comment by benedekrozemberczki** [comment]
Thank you @cthoyt!
